[Poll] Warning message on decryption failure of some events (PSG-1025)

mnaturel opened the pull request:
<!-- Please read [CONTRIBUTING.md](https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md) before submitting your pull request -->
 
## Type of change

- [x] Feature
- [ ] Bugfix
- [ ] Technical
- [ ] Other :

## Content

<!-- Describe shortly what has been changed -->
Making the encrypted events as processable. Processing the events with `relatesTo` field. When an event relates to an existing poll, it is saved into the related summary entity so that a warning can be displayed to the user.

## Motivation and context

<!-- Provide link to the corresponding issue if applicable or explain the context -->
Closes #7824 

## Screenshots / GIFs

<!-- Only if UI have been changed
You can use a table like this to show screenshots comparison.
Uncomment this markdown table below and edit the last line `|||`:
|copy screenshot of before here|copy screenshot of after here|
-->

<!--
|Before|After|
|-|-|
|||
 -->

<img src="https://user-images.githubusercontent.com/46314705/209679559-e573fee6-7fba-4685-9919-21408695beda.png" width=300 />


## Tests

<!-- Explain how you tested your development -->
The decryption failure is rather hard to test. The UI has been tested by mocking decryption failure.
We should ensure poll is working as before.

- Create open or closed poll in a room
- Vote on the poll
- Check the UI for the message is correct
- End the poll
- Check the UI for the message is correct

## Tested devices

- [ ] Physical
- [x] Emulator
- OS version(s): Android 11

## Checklist

<!-- Depending on the Pull Request content, it can be acceptable if some of the following checkboxes stay unchecked. -->

- [ ] Changes has been tested on an Android device or Android emulator with API 21
- [ ] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#accessibility
- [x] Pull request is based on the develop branch
- [x] Pull request includes a new file under ./changelog.d. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#changelog
- [x] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [x] You've made a self review of your PR
- [ ] If you have modified the screen flow, or added new screens to the application, you have updated the test [UiAllScreensSanityTest.allScreensTest()](https://github.com/vector-im/element-android/blob/main/vector/src/androidTest/java/im/vector/app/ui/UiAllScreensSanityTest.kt#L73)

BillCarsonFr reviewed (changes requested):
I think i'd prefer to modify the exisiting EventInsertEntity and EventInsertLiveObserver to handle the UTD. And pass them once to EventInsertLiveProcessor with a new method processUnableToDecryptEvent()

mnaturel commented:
> I think i'd prefer to modify the exisiting EventInsertEntity and EventInsertLiveObserver to handle the UTD. And pass them once to EventInsertLiveProcessor with a new method processUnableToDecryptEvent()

In fact, I retried to make it work with existing `EventInsertEntity` but I see problems depending on the solution:
- setting `canBeProcessed` field to `true` for encrypted events may lead to multiple processing of the same events. When the event is finally decrypted we need to [create or update the entity](https://github.com/vector-im/element-android/blob/5c442296efa532bd48a42b3ae80b7dbb6d517b9f/matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/database/model/EventEntity.kt#L98) but we cannot know whether the entity is currently beeing processed due to access of the database at the same time...
- when adding another observer using a filter on `canBeProcessed` to `false`, we may end up processing multiple times the same encrypted event as in that case we will not delete the entity after process since it is still needed for the other existing observer on `canBeProcessed` to `true`

These are the reasons I introduced a new `Entity` with a separate observer. Do you have an idea of a way to use `EventInsertEntity`?

mnaturel commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/database/query/EventEntityQueries.kt:
```diff
@@ -32,10 +31,9 @@ internal fun EventEntity.copyToRealmOrIgnore(realm: Realm, insertType: EventInse
             .equalTo(EventEntityFields.ROOM_ID, roomId)
             .findFirst()
     return if (eventEntity == null) {
-        val canBeProcessed = type != EventType.ENCRYPTED || decryptionResultJson != null
```
Is it possible to have a case where the type is not  `EventType.ENCRYPTED` and `decryptionResultJson` is set to `null`? If this case exists, I wonder if we should make it processable?

mnaturel replied:
Thanks for the explanation. It made me realize I needed to change some logic in `EventInsertLiveObserver`. I have updated it in [60e838a](https://github.com/vector-im/element-android/pull/7851/commits/60e838a82c4d82b26e1141f79cb7028bab50a7c6).

mnaturel commented:
> > I think i'd prefer to modify the exisiting EventInsertEntity and EventInsertLiveObserver to handle the UTD. And pass them once to EventInsertLiveProcessor with a new method processUnableToDecryptEvent()
> 
> In fact, I retried to make it work with existing `EventInsertEntity` but I see problems depending on the solution:
> 
>     * setting `canBeProcessed` field to `true` for encrypted events may lead to multiple processing of the same events. When the event is finally decrypted we need to [create or update the entity](https://github.com/vector-im/element-android/blob/5c442296efa532bd48a42b3ae80b7dbb6d517b9f/matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/database/model/EventEntity.kt#L98) but we cannot know whether the entity is currently beeing processed due to access of the database at the same time...
> 
>     * when adding another observer using a filter on `canBeProcessed` to `false`, we may end up processing multiple times the same encrypted event as in that case we will not delete the entity after process since it is still needed for the other existing observer on `canBeProcessed` to `true`
> 
> 
> These are the reasons I introduced a new `Entity` with a separate observer. Do you have an idea of a way to use `EventInsertEntity`?

I have reverted to use `EventInsertEntity` in [18ace4b](https://github.com/vector-im/element-android/pull/7851/commits/18ace4beb29d6a4e6d29807e22988722ab5392d1). Due to usage of a Realm transaction in the observer, it seems we won't have any race conditions issues.

BillCarsonFr commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/EventRelationsAggregationProcessor.kt:
```diff
@@ -73,6 +74,7 @@ internal class EventRelationsAggregationProcessor @Inject constructor(
+        private val encryptedReferenceAggregationProcessor: EncryptedReferenceAggregationProcessor,
         private val editValidator: EventEditValidator,
         private val clock: Clock,
 ) : EventInsertLiveProcessor {
```
I wonder if we should not update the `shouldProcess()` methods and for UTD events just look if there is a `m.relates_to` field? 
Actually it could even be done at the `EventInsertLiveObserver` level? not sure why we would like to observe UTD events that are not relations?

mnaturel replied:
That's a good point, I agree, I will update the `EventInsertLiveObserver`.

mnaturel replied:
Done in [ee3cbd9](https://github.com/vector-im/element-android/pull/7851/commits/ee3cbd988481d0d071726831c720389fd60c176a)

BillCarsonFr reviewed (changes requested):
Looks good, I just have 2 comments.

mnaturel commented on vector/src/main/java/im/vector/app/features/home/room/detail/timeline/item/MessageInformationData.kt:
```diff
@@ -90,7 +90,8 @@ data class PollResponseData(
         val winnerVoteCount: Int = 0,
-        val isClosed: Boolean = false
+        val isClosed: Boolean = false,
+        val hasDecryptionError: Boolean = false,
```
Yes exactly, I have renamed it to `hasEncryptedRelatedEvents` instead. See [d1ce15b](https://github.com/vector-im/element-android/pull/7851/commits/d1ce15bf18ea4119ece633dc4b42cf08cc5a3316)

BillCarsonFr reviewed (approved):
LGTM
